c-dependencies/js-compute-runtime/builtins/crypto-algorithm.cpp

```diff
@@ -183,6 +440,300 @@ std::unique_ptr<CryptoAlgorithmDigest> CryptoAlgorithmDigest::normalize(JSContex
+  // RSASSA-PKCS1-v1_5, RSA-PSS, RSA-OAEP, ECDSA, ECDH, AES-CTR, AES-CBC, AES-GCM, AES-KW, HMAC, HKDF, PBKDF2 are the algorithms 
+  // which support the importKey operation
+  switch (identifier) {
+  case CryptoAlgorithmIdentifier::RSASSA_PKCS1_v1_5: {
+    return CryptoAlgorithmRSASSA_PKCS1_v1_5_Import::fromParameters(cx, params);
+  }
+  case CryptoAlgorithmIdentifier::RSA_PSS:
+  case CryptoAlgorithmIdentifier::RSA_OAEP:
+  case CryptoAlgorithmIdentifier::AES_CTR:
+  case CryptoAlgorithmIdentifier::AES_CBC:
+  case CryptoAlgorithmIdentifier::AES_GCM:
+  case CryptoAlgorithmIdentifier::AES_KW:
+  case CryptoAlgorithmIdentifier::HMAC:
+  case CryptoAlgorithmIdentifier::ECDSA:
+  case CryptoAlgorithmIdentifier::ECDH:
+  case CryptoAlgorithmIdentifier::HKDF:
+  case CryptoAlgorithmIdentifier::PBKDF2: {
+    MOZ_ASSERT(false);
+    JS_ReportErrorASCII(cx, "Supplied algorithm is not yet supported");
+    convertErrorToNotSupported(cx);
+    return nullptr;
+  }
+  default: {
+    JS_ReportErrorASCII(cx, "Supplied algorithm does not support the importKey operation");
+    convertErrorToNotSupported(cx);
+    return nullptr;
+  }
+  }
+};
+
+std::unique_ptr<CryptoAlgorithmRSASSA_PKCS1_v1_5_Import> CryptoAlgorithmRSASSA_PKCS1_v1_5_Import::fromParameters(JSContext *cx, JS::HandleObject parameters) {
+  JS::Rooted<JS::Value> hash_val(cx);
+  if (!JS_GetProperty(cx, parameters, "hash", &hash_val)) {
+    return nullptr;
+  }
+  auto hashIdentifier = toHashIdentifier(cx, hash_val);
+  if (hashIdentifier.isErr()) {
+    return nullptr;
+  }
+  return std::make_unique<CryptoAlgorithmRSASSA_PKCS1_v1_5_Import>(hashIdentifier.unwrap());
+}
+
+// https://w3c.github.io/webcrypto/#rsassa-pkcs1-operations
+JSObject *CryptoAlgorithmRSASSA_PKCS1_v1_5_Import::importKey(JSContext *cx, CryptoKeyFormat format,
+                                                             KeyData keyData, bool extractable,
+                                                             CryptoKeyUsages usages) {
+  MOZ_ASSERT(cx);
+  JS::RootedObject result(cx);
+  switch (format) {
+  // 2. If format is "jwk":
+  case CryptoKeyFormat::Jwk: {
+    // 2.1. If keyData is a JsonWebKey dictionary:
+      // Let jwk equal keyData.
+    // Otherwise:
+      // Throw a DataError.
+    auto jwk = std::get<JsonWebKey *>(keyData);
+    if (!jwk) {
+      // TODO Rename error to DataError
+      JS_ReportErrorLatin1(cx, "Supplied format is not a JSONWebKey");
+      return nullptr;
+    }
+
+
+    // 2.2 If the d field of jwk is present and usages contains an entry which 
+    // is not "sign", or, if the d field of jwk is not present and usages 
+    // contains an entry which is not "verify" then throw a SyntaxError.
+    bool isUsagesAllowed = false;
+    // public key
+    if (jwk->d.has_value()) {
+      isUsagesAllowed = usages.canSign();
+    } else {
+      // private key
+      isUsagesAllowed = usages.canVerify();
+    }
+    isUsagesAllowed = isUsagesAllowed || !usages.isEmpty();
```

Comment:
Why would usages be allowed if `usages` was not empty? The way I'm reading the comment it seems like the two cases are just those of the conditional above.